File: androidApp/src/main/java/com/app/pustakam/android/screen/noteEditor/NoteEditorViewModel.kt
package com.app.pustakam.android.screen.noteEditor

import com.app.pustakam.core.common.util.displayMessage
import android.annotation.SuppressLint
import android.content.Context
import androidx.compose.runtime.mutableStateListOf
import androidx.lifecycle.viewModelScope
import com.app.pustakam.android.fileUtils.deleteFile
import com.app.pustakam.android.fileUtils.generateThumbnail
import com.app.pustakam.android.noteContentProvider.addContent
import com.app.pustakam.android.permission.NeededPermission
import com.app.pustakam.android.screen.NOTES_CODES
import com.app.pustakam.android.screen.NoteContentUiState
import com.app.pustakam.android.screen.NoteUIState
import com.app.pustakam.android.screen.TaskCode
import com.app.pustakam.android.screen.base.BaseViewModel
import com.app.pustakam.feature.notes.domain.usecase.CreateORUpdateNoteUseCase
import com.app.pustakam.feature.notes.domain.usecase.DeleteNoteContentUseCase
import com.app.pustakam.feature.notes.domain.usecase.DeleteNoteUseCase
import com.app.pustakam.feature.notes.domain.usecase.ReadNoteUseCase
import com.app.pustakam.core.model.models.BaseResponse
import com.app.pustakam.core.model.models.response.notes.Note
import com.app.pustakam.core.model.models.response.notes.NoteContentModel
import com.app.pustakam.core.model.models.response.notes.NoteContentObjectHelper
import com.app.pustakam.core.model.models.response.notes.TextBlockSplitter
import com.app.pustakam.feature.notes.domain.usecase.ClearSelectedNoteContentUseCase
import com.app.pustakam.feature.notes.domain.usecase.SetSelectedNoteContentUseCase
import com.app.pustakam.feature.notes.domain.usecase.UpdateSelectedMediaContentUseCase
import com.app.pustakam.core.common.extensions.isNotnull
import com.app.pustakam.core.common.util.ContentType
import com.app.pustakam.core.common.util.ContentType.AUDIO
import com.app.pustakam.core.common.util.ContentType.IMAGE
import com.app.pustakam.core.common.util.ContentType.LOCATION
import com.app.pustakam.core.common.util.ContentType.TEXT
import com.app.pustakam.core.common.util.ContentType.VIDEO
import com.app.pustakam.core.common.util.Error
import com.app.pustakam.core.common.util.NetworkError
import com.app.pustakam.core.common.util.log_d
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.core.component.inject
import com.app.pustakam.core.common.util.Result

class NoteEditorViewModel : BaseViewModel() {
    private val setSelectedNoteContentUseCase by inject<SetSelectedNoteContentUseCase>()
    private val updateSelectedMediaContentUseCase by inject<UpdateSelectedMediaContentUseCase>()
    private val clearSelectedNoteContentUseCase by inject<ClearSelectedNoteContentUseCase>()
    private val readNoteUseCase by inject<ReadNoteUseCase>()
    private val deleteNoteUseCase by inject<DeleteNoteUseCase>()
    private val deleteNoteContentUseCase by inject<DeleteNoteContentUseCase>()
    private val createUpdateNoteUseCase by inject<CreateORUpdateNoteUseCase>()
    private val _noteUiState = MutableStateFlow(NoteUIState(isLoading = false))
    val noteUIState: StateFlow<NoteUIState> = _noteUiState.asStateFlow()
    private val _noteContentUiState = MutableStateFlow(NoteContentUiState())
    val noteContentUiState: StateFlow<NoteContentUiState> = _noteContentUiState.asStateFlow()
    private val dirtyContentIds = mutableSetOf<String>()

    init {
        viewModelScope.launch {
            setSelectedNoteContentUseCase.selectedMediaContent.collect { media ->
                media.forEach { applyExternalContentUpdate(it) }
            }
        }
    }

    private fun applyExternalContentUpdate(updated: NoteContentModel) {
        _noteContentUiState.update { state ->
            val index = state.contents.indexOfFirst { it.id == updated.id }
            if (index == -1) return@update state
            if (state.contents[index].updatedAt == updated.updatedAt) return@update state
            state.contents[index] = updated
            val updatedNote = state.note?.let { n ->
                n.withContents(n.contents.map { if (it.id == updated.id) updated else it })
            }
            state.copy(note = updatedNote, contents = state.contents)
        }
    }

    private var lastSavedDirtyIds: Set<String> = emptySet()


    //default methods
    override fun onLoading(taskCode: TaskCode) {
        _noteUiState.update {
            it.copy(isLoading = true, successMessage = "")
        }
    }

    override fun onSuccess(taskCode: TaskCode, result: Result.Success<BaseResponse<*>>) {
        when (taskCode) {
            NOTES_CODES.INSERT -> {
                log_d("Loading", "Getting Update data")
                val note = result.data.data as Note
                _noteUiState.update {
                    val noteStatus = when (it.noteStatus) {
                        NoteStatus.onBackPress, NoteStatus.onSaveCompletedExit -> NoteStatus.onSaveCompletedExit
                        else -> NoteStatus.onSaveCompleted
                    }
                    it.copy(noteStatus = noteStatus, isLoading = false)
                }

                _noteContentUiState.update { it.copy(note = note,
                    isAllSetupDone = true) }

                dirtyContentIds.removeAll(lastSavedDirtyIds)
                lastSavedDirtyIds = emptySet()
                consumePendingMediaPaths()
            }

            NOTES_CODES.READ -> {
                val note = result.data.data as Note
                if (!noteContentUiState.value.isAllSetupDone) _noteContentUiState.update {
                    it.titleTextState.value = note.title ?: ""
                    it.copy(
                        titleTextState = it.titleTextState, note = note,
                        isAllSetupDone = true,

                        contents = mutableStateListOf(*note.contents.sortedBy { c -> c.position }.toTypedArray())
                    )
                }

                setSelectedNoteContentUseCase(_noteContentUiState.value.note ?: note)

                consumePendingMediaPaths()
            }

            NOTES_CODES.DELETE -> {
                _noteUiState.update {
                    it.copy(
                        isLoading = false, noteStatus = NoteStatus.exit
                    )
                }
            }
        }
    }

    override fun onFailure(taskCode: TaskCode, error: Error) {
        when (taskCode) {
            NOTES_CODES.READ -> {
                _noteUiState.update {
                    it.copy(
                        isLoading = false, error = error.displayMessage(), noteStatus = NoteStatus.onSaveCompletedExit
                    )
                }
            }

            else -> _noteUiState.update {
                it.copy(isLoading = false, error = error.displayMessage(), noteStatus = null)
            }
        }
    }


    override fun clearError() {
        _noteUiState.update {
            it.copy(
                error = null, successMessage = null, isLoading = false
            )
        }
    }

    /** CRUD operations on Notes */

    // call make a wish api
    fun isNoteValid(): Boolean =
        if (_noteContentUiState.value.titleTextState.value.isEmpty()) {
            if (_noteContentUiState.value.note?.contents?.isEmpty() == true) {
                 false
            } else  true
        } else true

    fun createOrUpdateNote() {
        if(!isNoteValid()) {
                changeNoteStatus(NoteStatus.exit)
                return
        }
        updateNoteObject()
        // 🔧 15-Jul-2026 Phase 0.4: snapshot the dirty ids for THIS save; cleared on INSERT success.
        lastSavedDirtyIds = dirtyContentIds.toSet()
        makeAWish(NOTES_CODES.INSERT) {
            createUpdateNoteUseCase.invoke(_noteContentUiState.value.note!!, lastSavedDirtyIds)
        }
    }
    fun saveThenOpen(onSaved: () -> Unit) {
        updateNoteObject()
        val note = _noteContentUiState.value.note ?: run { onSaved(); return }
        val dirty = dirtyContentIds.toSet()
        viewModelScope.launch(Dispatchers.IO) {
            createUpdateNoteUseCase.invoke(note, dirty).collect { result ->
                if (result is Result.Success || result is Result.Error) {
                    dirtyContentIds.removeAll(dirty)
                    withContext(Dispatchers.Main) { onSaved() }
                }
            }
        }
    }

    fun readFromDataBase(id: String?) {
        if (!id.isNullOrEmpty()) _noteUiState.update {
            it.copy(showDeleteButton = true)
        }
        makeAWish(NOTES_CODES.READ) {
            readNoteUseCase.invoke(id)
        }
    }

    fun refreshOnResume(id: String?) {
        if (dirtyContentIds.isNotEmpty()) return   // don't overwrite unsaved edits
        readFromDataBase(id)
    }

    fun deleteNote(noteId: String) {
        makeAWish(NOTES_CODES.DELETE) {
            deleteNoteUseCase.invoke(noteId)
        }
    }

    private fun updateNoteObject() {
        TextBlockSplitter.splitOversized(_noteContentUiState.value.contents.toList())?.let { split ->
            dirtyContentIds.addAll(split.changedIds)
            val live = _noteContentUiState.value.contents
            live.clear()
            live.addAll(split.contents)
        }
        _noteContentUiState.update {
            val updatedNote = it.note?.withTitleAndContents(
                newTitle = it.titleTextState.value,
                newContents = it.contents.toList()
            )
            it.copy(note = updatedNote)
        }
    }


    //action status methods
    fun changeNoteStatus(status: NoteStatus?) {
        _noteUiState.update {
            it.copy(
                noteStatus = status, isLoading = false
            )
        }
    }

    private fun setContentType(contentType: ContentType) {
        _noteUiState.update {
            it.copy(contentType = contentType, isLoading = true)
        }
    }

    //dialog trigger methods
    fun showDeleteAlertBox(value: Boolean, deleteNoteContentId: String? = null) {
        // some time it doesn't update a single value
        _noteUiState.update { it.copy(showDeleteAlert = value, deleteNoteContentId = deleteNoteContentId, isLoading = false) }
    }

    fun showPermissionAlert(value: Boolean?) {
        _noteUiState.update {
            it.copy(showPermissionAlert = value)
        }
    }

    //hardware permission logic
    @SuppressLint("NewApi")
    private fun getPermissions(contentType: ContentType?) = when (contentType) {
        VIDEO -> listOf(NeededPermission.CAMERA, NeededPermission.RECORD_AUDIO)
        AUDIO -> listOf(NeededPermission.RECORD_AUDIO)
        IMAGE -> listOf(NeededPermission.CAMERA)
        LOCATION -> listOf(NeededPermission.COARSE_LOCATION, NeededPermission.FINE_LOCATION,
            NeededPermission.BACKGROUND_LOCATION)
        else -> listOf(NeededPermission.POST_NOTIFICATIONS)
    }

   /**permission dialog setup*/
    fun preparePermissionDialog(contentType: ContentType? = null) {
        val permission = getPermissions(contentType)
        _noteUiState.update {
            it.copy(permissions = permission, contentType = contentType, showPermissionAlert = true, isLoading = false)
        }
    }
      fun addNewText() {
          val note = _noteContentUiState.value.note!!
          if (note.isNotnull()) {
              val textContent =
                  NoteContentObjectHelper.createText(
                      noteId = note.id,
                      positionedAt = note.contents.count().toDouble()   // 🔧 C4
                  )
              setContentType(TEXT)
              updateContent(content = textContent)
          }
      }

    fun locationState(value: Boolean ) {
        _noteUiState.update { it.copy(LocationState = value) }
    }
    fun updateContent(index: Int = -1, content: NoteContentModel) {
        dirtyContentIds.add(content.id)   // 🔧 15-Jul-2026 Phase 0.4: touched → will be saved
        if(index== -1) {
            addContentData(content)
        }else{
            _noteContentUiState.update {
                it.contents[index] = content
                // immutable Note: upsert by id via copy (was: in-place add — also fixed
                // the old bug of ADDING a duplicate on update instead of replacing)
                val updatedNote = it.note?.let { n ->
                    val newContents = n.contents.toMutableList()
                    val i = newContents.indexOfFirst { c -> c.id == content.id }
                    if (i != -1) newContents[i] = content else newContents.add(content)
                    n.withContents(newContents)
                }
                it.copy(note = updatedNote, contents = it.contents, isAllSetupDone = true)
            }
        }
        if (content.isPlayingMedia())
            updateSelectedMediaContentUseCase(content as NoteContentModel.MediaContent)
    }
    /**content logic
     * Add new content to the note content list
     * by selecting it type on the bases of user selection
     * */
    fun addNewContent(context: Context, contentType: ContentType): NoteContentModel {
        setContentType(contentType)
        val content = addContent( context = context, note = _noteContentUiState.value.note!!, contentType = contentType)
        return content
    }
    fun addContentData(content: NoteContentModel){
        dirtyContentIds.add(content.id)

        val liveContents = _noteContentUiState.value.contents
        val existingIndex = liveContents.indexOfFirst { it.id == content.id }
        if (existingIndex != -1) liveContents[existingIndex] = content else liveContents.add(content)
        _noteContentUiState.update {
            val updatedNote = it.note?.let { n ->
                val newContents = n.contents.toMutableList()
                val i = newContents.indexOfFirst { c -> c.id == content.id }
                if (i != -1) newContents[i] = content else newContents.add(content)
                n.withContents(newContents)
            }
            it.copy(note = updatedNote, contents = it.contents, isAllSetupDone = true)
        }
    }

    fun startStopAudioRecording(value: Boolean = true) {
        _noteUiState.update { it.copy(isLoading = false, showAudioRecorder = value) }
    }
/**
 * Remove a note content for note
 * */
    fun removeContent(value: String) {
        dirtyContentIds.remove(value)   // 🔧 15-Jul-2026 Phase 0.4: deleted → nothing to save
        val find = _noteContentUiState.value.note?.contents?.find { value == it.id }
        viewModelScope.launch(Dispatchers.IO) {
            var rowDeleted = false
            deleteNoteContentUseCase.invoke(value).collect { result ->
                when (result) {
                    is Result.Success -> rowDeleted = true
                    is Result.Error -> log_d("NoteEditor", "content delete failed: ${result.error}")
                    else -> Unit
                }
            }

            if (rowDeleted && find?.isMediaFile() == true) {
                find as NoteContentModel.MediaContent
                find.localPath?.let { deleteFile(filePath = it) }
                // 🔧 15-Jul-2026 (iOS-parity cleanup): the media's thumbnail file goes with it
                find.thumbnailPath?.let { deleteFile(filePath = it) }
            }
        }
        _noteContentUiState.update {
            val indexContent = it.contents.indexOf(find)
            if (indexContent != -1) it.contents.removeAt(indexContent)
            // immutable Note: rebuild contents without the removed item.
            // (old code removed from a discarded copy — note.contents was never
            //  actually updated; this also fixes that silent bug)
            val updatedNote = it.note?.let { n ->
                n.withContents(n.contents.filterNot { c -> c.id == value })
            }
            it.copy(note = updatedNote, contents = it.contents)
        }
        showDeleteAlertBox(false, null)
    }


 /**
  * Share note with others
  */
    fun shareNote(){}

    override fun onCleared() {
        super.onCleared()
        clearSelectedNoteContentUseCase()
    }

    // 🔧 14-Jul-2026: CRASH FIX + PERF — this used to run inside composition and mutate
    //   the snapshot list INSIDE StateFlow.update{} (which can re-run its lambda), producing
    //   duplicate items and, with several images at once, a crash. Now it is pure: we build
    //   ALL new MediaContent items first, mutate the SnapshotStateList exactly once (outside
    //   the update lambda), then do a single state copy. Callers invoke it from a LaunchedEffect.
    //
    // 🔧 14-Jul-2026: FIX (I-1, video lost after Stop→Back) — the old `note ?: return` silently
    //   DROPPED the captured paths when the note hadn't loaded yet (the View clears them right
    //   after this call). Paths arriving too early are now stashed in `pendingMediaPaths` and
    //   consumed as soon as READ/INSERT delivers the note — nothing is ever lost.
    private var pendingMediaPaths: List<Pair<String, ContentType>> = emptyList()

    private fun consumePendingMediaPaths() {
        if (pendingMediaPaths.isEmpty()) return
        val pending = pendingMediaPaths
        pendingMediaPaths = emptyList()
        getMediaData(pending)
    }

    // 🔧 15-Jul-2026 Phase 2.3: application context for async thumbnail generation (safe to hold —
    //   never an Activity). Set by getMediaData; used again when pending paths are consumed.
    private var appContext: Context? = null

    fun getMediaData(list: List<Pair<String, ContentType>>, context: Context? = null) {
        if (context != null) appContext = context.applicationContext
        if (list.isEmpty()) return
        val currentState = _noteContentUiState.value
        val note = currentState.note ?: run {
            pendingMediaPaths = pendingMediaPaths + list   // 🔧 stash instead of dropping
            return
        }
        var position: Double = note.contents.count().toDouble()   // 🔧 C4
        val newItems = list.map { path ->
            val content = NoteContentObjectHelper.createMedia(
                positionedAt = position,
                noteId = note.id,
                localPath = path.first,
                contentType = path.second
            ).copy(title = "${path.second}-$position")
            position += 1.0
            content
        }
        // Single mutation of the observed list (safe: called off the composition pass).
        currentState.contents.addAll(newItems)
        _noteContentUiState.update {
            it.copy(
                note = note.withContents(note.contents + newItems),
                contents = it.contents,
                isAllSetupDone = true
            )
        }
        newItems.filter { it.isPlayingMedia() }.forEach { updateSelectedMediaContentUseCase(it) }

        dirtyContentIds.addAll(newItems.map { it.id })
        generateThumbnailsFor(newItems)
    }

    fun importDeviceFiles(context: Context, uris: List<android.net.Uri>) {
        if (uris.isEmpty()) return
        appContext = context.applicationContext
        val note = _noteContentUiState.value.note ?: run {
            _noteUiState.update { it.copy(error = "Note is still loading. Try again.") }; return
        }
        _noteUiState.update { it.copy(isLoading = true) }
        viewModelScope.launch(Dispatchers.IO) {
            val items = com.app.pustakam.android.fileimport.FileImportManager.importUris(
                context.applicationContext, note.id, note.contents.count().toDouble(), uris
            )
           withContext(Dispatchers.Main) {
                _noteUiState.update { it.copy(isLoading = false) }
                if (items.isEmpty()) _noteUiState.update { it.copy(error = "Couldn't import the selected files.") }
                else addImportedContents(items)
            }
        }
    }


    fun importFromLink(context: Context, url: String) {
        if (url.isBlank()) return
        appContext = context.applicationContext
        val note = _noteContentUiState.value.note ?: run {
            _noteUiState.update { it.copy(error = "Note is still loading. Try again.") }; return
        }
        _noteUiState.update { it.copy(isLoading = true) }
        viewModelScope.launch(Dispatchers.IO) {
            val result = com.app.pustakam.android.fileimport.FileImportManager.importFromUrl(
                context.applicationContext, note.id, note.contents.count().toDouble(), url
            )
            withContext(Dispatchers.Main) {
                _noteUiState.update { it.copy(isLoading = false) }
                when (result) {
                    is com.app.pustakam.android.fileimport.ImportResult.Success -> addImportedContents(result.contents)
                    is com.app.pustakam.android.fileimport.ImportResult.NoFileFound ->
                        _noteUiState.update { it.copy(error = "No file found at this link.") }
                    is com.app.pustakam.android.fileimport.ImportResult.Failed ->
                        _noteUiState.update { it.copy(error = result.message) }
                }
            }
        }
    }

    // 🔧 18-Jul-2026: same safe mutation pattern as getMediaData — list touched ONCE outside update{}
    private fun addImportedContents(items: List<NoteContentModel.MediaContent>) {
        if (items.isEmpty()) return
        val currentState = _noteContentUiState.value
        val note = currentState.note ?: return
        currentState.contents.addAll(items)
        _noteContentUiState.update {
            it.copy(note = note.withContents(note.contents + items), contents = it.contents, isAllSetupDone = true)
        }
        items.filter { it.isPlayingMedia() }.forEach { updateSelectedMediaContentUseCase(it) }
        dirtyContentIds.addAll(items.map { it.id })   // 🔧 imported blocks are new rows → saved next save
        generateThumbnailsFor(items)
    }
    private fun generateThumbnailsFor(items: List<NoteContentModel.MediaContent>) {
        val context = appContext ?: return
        items.filter {
            it.thumbnailPath.isNullOrEmpty() && !it.localPath.isNullOrEmpty() &&
                    (it.type == VIDEO || it.type == IMAGE || it.type == ContentType.GIF)
        }.forEach { media ->
            viewModelScope.launch(Dispatchers.IO) {
                val thumb = generateThumbnail(context, media.localPath!!, media.type) ?: return@launch
                val latest = _noteContentUiState.value.contents
                    .filterIsInstance<NoteContentModel.MediaContent>()
                    .firstOrNull { it.id == media.id } ?: media
                updateContent(content = latest.copy(thumbnailPath = thumb))
            }
        }
    }
}

File: iosApp/iosApp/view/screens/notes/noteEditor/NoteEditorViewModel.swift
import shared
import Combine
import SwiftUI

/// Single UI-state surface for the editor (mirror of the list slice pattern).
struct NoteEditorUIState {
    var note: Note? = nil                       // fixes E2: now inside @Published state
    var title: String = ""                      // fixes V2: title owned by VM, saved reliably
    var noteContents: [NoteContentModel] = []   // SINGLE source of truth (fixes E4)
    var isLoading = false
    var isDeleted = false                       // fixes V3: save() refuses after delete
    var errorMessage: String? = nil
    var isNoteReady: Bool { note != nil }
}

class NoteEditorViewModel: ObservableObject {

    @Published var state = NoteEditorUIState()
    @Published private(set) var capabilities = EditorCapabilityCommands.shared.empty()

    private let adapter: NotesBridgeAdapter
    private let contentBridge: NoteContentBridge
    private var contentUpdatesHandle: Closeable?
    private var dirtyContentIds = Set<String>()

    @Published private(set) var history = NoteHistory(
        past: [],
        future: [],
        limit: NoteHistory.companion.DEFAULT_LIMIT
    )

    var canUndo: Bool { history.canUndo }

    var canRedo: Bool { history.canRedo }

    private func currentSnapshot() -> Note? {
        guard let note = state.note else { return nil }
        return note.withTitleAndContents(newTitle: state.title, newContents: state.noteContents)
    }

    private func restore(_ note: Note) {
        state.note = note
        state.title = note.title ?? ""
        state.noteContents = note.contents as? [NoteContentModel] ?? []
        state.noteContents.forEach { dirtyContentIds.insert($0.id) }
        contentBridge.setSelectedNote(note: note)
    }

    func undo() {
        guard let current = currentSnapshot(),
              let step = history.undoStep(current: current) else { return }
        history = step.history
        restore(step.note)
    }

    func redo() {
        guard let current = currentSnapshot(),
              let step = history.redoStep(current: current) else { return }
        history = step.history
        restore(step.note)
    }

    /// DI per series convention: defaults keep call sites/tests simple. (fixes E1, E6)
    init(noteId: String? = nil,
         adapter: NotesBridgeAdapter = NotesBridgeAdapter(),
         contentBridge: NoteContentBridge = NoteContentBridge()) {
        self.adapter = adapter
        self.contentBridge = contentBridge
        load(noteId: noteId)
        observeContentUpdates()
    }

    deinit {
        contentUpdatesHandle?.close()
        contentBridge.dispose()
    }

    private func observeContentUpdates() {
        contentUpdatesHandle = contentBridge.observeSelectedMedia { [weak self] media in
            guard let self else { return }
            for updated in media {
                guard let index = self.state.noteContents.firstIndex(where: { $0.id == updated.id }) else { continue }
                guard self.state.noteContents[index].updatedAt != updated.updatedAt else { continue }
                self.state.noteContents[index] = updated
            }
        }
    }

    // MARK: - Load
    func refresh() {
        guard dirtyContentIds.isEmpty else { return }   // don't overwrite unsaved edits
        guard let id = state.note?.id, !id.isEmpty else { return }
        load(noteId: id)
    }

    private func load(noteId: String?) {
        adapter.readNote(noteId: noteId) { [weak self] result in
            guard let self else { return }
            switch result {
            case .loading:            self.state.isLoading = true
            case .success(let note):  self.state.isLoading = false
                                      if let note { self.apply(note: note) }
            case .failure(let error): self.state.isLoading = false
                                      self.state.errorMessage = error.message
            case .idle: break
            }
        }
    }

    private func apply(note: Note) {
        state.note = note
        state.title = note.title ?? ""
        state.noteContents = note.contents as? [NoteContentModel] ?? []
        contentBridge.setSelectedNote(note: note)
    }

    // MARK: - Content editing (single list; note.contents materialized only at save)

    func addContent(content: NoteContentModel) {
        guard state.isNoteReady else { return }                 // fixes E3 crash window
        dirtyContentIds.insert(content.id)   // 🔧 15-Jul-2026 iOS parity: new block → must be saved
        state.noteContents.append(content)
        if content.isPlayingMedia(), let media = content as? NoteContentModel.MediaContent {
            contentBridge.updateMediaContent(content: media)    // safe cast (was as!)
        }
    }

    func updateContent(content: NoteContentModel) {
        if let snapshot = currentSnapshot() {
            history = content is NoteContentModel.TextContent
                ? history.recordText(previous: snapshot)
                : history.recordAddMedia(previous: snapshot)
        }
        dirtyContentIds.insert(content.id)   // 🔧 15-Jul-2026 iOS parity: touched → will be saved
        if let index = state.noteContents.firstIndex(where: { $0.id == content.id }) {
            state.noteContents[index] = content
        } else {
            addContent(content: content)
        }
        // NOTE: no more parallel note.contents bookkeeping (E4) — save() materializes.
    }

    func addNewText() {
        if let snapshot = currentSnapshot() { history = history.recordAddText(previous: snapshot) }
        guard let noteId = state.note?.id else { return }       // fixes E3 (was note!.id)
        let text = NoteContentObjectHelper.shared.createText(
            noteId: noteId,
            positionedAt: Double(state.noteContents.count),     // 🔧 C4: position is Double now
            text: "")
        addContent(content: text)
    }

    func getCapturedData(media: CapturedMedia?) {
        guard let noteId = state.note?.id else { return }
        guard let content = EditorCapture.persist(
            media: media,
            noteId: noteId,
            positionedAt: Double(state.noteContents.count)
        ) else { return }
        addContent(content: content)
        generateThumbnailAsync(for: content)
        capabilities = EditorCapabilityReducer.shared.captureFinished(state: capabilities)
    }

    func onCapabilityState(_ next: EditorCapabilityState) {
        capabilities = next
    }

    func requestCapture(_ kind: CaptureKind) {
        capabilities = EditorCapabilityReducer.shared.requestCapture(
            state: capabilities,
            kind: kind
        )
    }

    func askDeleteContent(_ contentId: String) {
        capabilities = EditorCapabilityReducer.shared.askDeleteContent(
            state: capabilities,
            contentId: contentId
        )
    }

    func askDeleteNote() {
        capabilities = EditorCapabilityReducer.shared.askDeleteNote(state: capabilities)
    }

    // MARK: - File import (18-Jul-2026)

    // 🔧 18-Jul-2026: NEW FEATURE (file import) — device multi-pick: copy on background, append on main
    func importFiles(urls: [URL]) {
        guard let noteId = state.note?.id else {
            state.errorMessage = "Note is still loading. Try again."
            return
        }
        guard !urls.isEmpty else { return }
        state.isLoading = true
        let startPosition = Double(state.noteContents.count)
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let items = FileImportService.importPicked(urls: urls, noteId: noteId, startPosition: startPosition)
            DispatchQueue.main.async {
                guard let self else { return }
                self.state.isLoading = false
                guard !items.isEmpty else {
                    self.state.errorMessage = "Couldn't import the selected files."
                    return
                }
                items.forEach { self.addContent(content: $0) }          // dirty + playlist handled inside
                items.forEach { self.generateThumbnailAsync(for: $0) }  // image/video thumbs only
            }
        }
    }

    // 🔧 18-Jul-2026: NEW FEATURE (file import) — link import: download if it IS a file, else "No file found"
    func importFromLink(_ url: String) {
        guard let noteId = state.note?.id else {
            state.errorMessage = "Note is still loading. Try again."
            return
        }
        guard !url.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        state.isLoading = true
        FileImportService.importFromLink(url, noteId: noteId,
                                         startPosition: Double(state.noteContents.count)) { [weak self] result in
            guard let self else { return }
            self.state.isLoading = false
            switch result {
            case .success(let items):
                items.forEach { self.addContent(content: $0) }
                items.forEach { self.generateThumbnailAsync(for: $0) }
            case .noFileFound:
                self.state.errorMessage = "No file found at this link."
            case .failed(let message):
                self.state.errorMessage = message
            }
        }
    }

    private func generateThumbnailAsync(for media: NoteContentModel.MediaContent) {
        guard media.type == ContentType.image || media.type == ContentType.video,
              let path = media.localPath, !path.isEmpty else { return }
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let thumb = generateThumbnail(sourcePath: path, type: media.type) else { return }
            DispatchQueue.main.async {
                guard let self else { return }
                let latest = self.state.noteContents.first { $0.id == media.id }
                    as? NoteContentModel.MediaContent ?? media
                self.updateContent(content: latest.withThumbnail(path: thumb))
            }
        }
    }

    // MARK: - Save / Delete
    func saveThenOpen(onSaved: @escaping () -> Void) {
        guard !state.isDeleted, let note = state.note, !state.noteContents.isEmpty else {
            onSaved(); return
        }
        let toSave = note.withTitleAndContents(newTitle: state.title, newContents: state.noteContents)
        let dirtySnapshot = dirtyContentIds
        adapter.createOrUpdateNote(note: toSave, dirtyContentIds: dirtySnapshot) { [weak self] result in
            switch result {
            case .success:
                self?.dirtyContentIds.subtract(dirtySnapshot)
                onSaved()
            case .failure(let error):
                // still navigate — the reader will show its own error if the file truly can't load
                print("saveThenOpen failed [\(error.code)] \(error.message)")
                onSaved()
            case .loading, .idle:
                break
            }
        }
    }
    func isNoteValid() -> Bool {
        guard let note = state.note else { return false}
        if (state.title.isEmpty) {
            if (state.noteContents.isEmpty){
               return false
            }
        }
        return true
    }


    func saveNote() {
         guard !state.isDeleted, let note = state.note, isNoteValid() else { return }
        
        if let split = TextBlockSplitter.shared.splitOversized(contents: state.noteContents) {
            state.noteContents = split.contents as? [NoteContentModel] ?? state.noteContents
            split.changedIds.forEach { id in
                
                if let id = id as? String { dirtyContentIds.insert(id) }
            }
        }
        let toSave = note.withTitleAndContents(
            newTitle: state.title,
            newContents: state.noteContents
        )

        let dirtySnapshot = dirtyContentIds
        adapter.createOrUpdateNote(note: toSave, dirtyContentIds: dirtySnapshot) { [weak self] result in
            switch result {
            case .success:
                self?.dirtyContentIds.subtract(dirtySnapshot)
            case .failure(let error):
                self?.state.errorMessage = error.message
                print("saveNote failed [\(error.code)] \(error.message)")
            case .loading, .idle:
                break
            }
        }
    }

    /// Replaces deleteNoteCall(); View no longer runs Task/casts. (fixes V5)
    func deleteNote(onDeleted: @escaping () -> Void) {
        guard let noteId = state.note?.id else { return }
        adapter.deleteNote(noteId: noteId) { [weak self] result in
            guard let self else { return }
            switch result {
            case .loading:
                self.state.isLoading = true
            case .success:
                self.state.isLoading = false
                self.state.isDeleted = true
                onDelete(noteId: noteId)
                onDeleted()
                // View dismisses
            case .failure(let error):
                self.state.isLoading = false
                self.state.errorMessage = error.message
            case .idle: break
            }
        }
    }
   private func onDelete(noteId : String) {
        adapter.deleteNote(noteId: noteId , onState: { _ in })
    }
    func deleteContent(contentId: String) {
        if let snapshot = currentSnapshot() {
            history = history.recordDeleteContent(previous: snapshot)
        }
        dirtyContentIds.remove(contentId)   // 🔧 15-Jul-2026 iOS parity: deleted → nothing to save
        guard let content = state.noteContents.first(where: { $0.id == contentId }) else { return }

        // 1. remove local media file (image/video/audio) — safe no-op for text
        if content.isMediaFile(), let media = content as? NoteContentModel.MediaContent,
           let localPath = media.localPath {
            deleteFile(filePath: LocalFilePathResolver_iosKt.resolveLocalFilePath(path: localPath) ?? localPath)
            // thumbnail lives in Documents/thumbnails — remove it with its media
            if let thumb = LocalFilePathResolver_iosKt.resolveLocalFilePath(path: media.thumbnailPath) {
                deleteFile(filePath: thumb)
            }
        }

        // 2. remove from DB (write call — survives screen death, like saveNote)
        adapter.deleteNoteContent(contentId: contentId) { [weak self] result in
            if case .failure(let error) = result {
                self?.state.errorMessage = error.message
                print("deleteContent failed [\(error.code)] \(error.message)")
            }
        }

        // 3. remove from UI state — save() materializes contents from this list,
        //    so the deleted block can never come back on the next save
        state.noteContents.removeAll { $0.id == contentId }
    }

    func shareNote() {}     // stub kept (nothing deleted)
}
